File: Top_100_Liked_Questions/240._Search_a_2D_Matrix_II/solution.py
class Solution:
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        # Search space reduction method
        numRow = len(matrix)
        if not numRow:
            return False
        
        numCol = len(matrix[0])
        if not numCol:
            return False
        
        r, c = numRow-1, 0
        while r >= 0 and c <= numCol-1:
            if matrix[r][c] < target:
                c += 1
            elif matrix[r][c] > target:
                r -= 1
            else:
                return True
        
        return False

File: Top_100_Liked_Questions/240._Search_a_2D_Matrix_II/solution.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int numRow = matrix.length;
        if (numRow == 0)
            return false;

        int numCol = matrix[0].length;
        if (numCol == 0)
            return false;

        int row = numRow-1;
        int col = 0;

        while (true) {
            if (matrix[row][col] < target) {
                if (++col == numCol)
                    return false;
            } else if (matrix[row][col] > target) {
                if (row-- == 0)
                    return false;
            } else
                return true;
        }
    }
}